### DISCUSSION TITLE
Waiting for provisioned concurrency on Lambdas

### DISCUSSION THREAD

**Discussion by NikolayMetchev** [aws/aws-sdk-kotlin#1350] [Q&A] [answered]
If you have an AWS Lambda with provisioned concurrency on an alias and you update the alias. The effect isn't immediate. It takes time for the provisioned concurrency to be updated for the Lambda alias. Is there a code example of how to wait for the provisioned concurrency to be updated before proceeding.

**Comment by lauzadis** [comment]
Does updating the alias transition the Lambda function's `LastUpdateStatus` to "InProgress"? If so, you can use the waiter function `LambdaClient.waitUntilFunctionUpdatedV2(...)` to wait for the `LastUpdateStatus` to transition to "Successful".

Here is our developer guide entry on waiters: https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/waiters.html

**Reply by NikolayMetchev** [reply]
Nope, it doesn't. There is a separate status for provisioned concurrency

**Reply by lauzadis** [reply]
Oh ok. I would still recommend trying the waiter function, it should wait until the status is "Successful"

**Reply by NikolayMetchev** [reply]
As far as I can tell the waiter function only checks lastUpdateStatus which is "Successful" even if ProvisionedConcurrencyStatus is Pending.

**Reply by NikolayMetchev** [reply]
In fact that makes sense because the Provisioned Concurrency Status is not on a function itself but rather on a version/alias so it can't be part of the function update.

**Reply by ianbotsf** [reply]
I don't think a waiter exists for this. If it did, it would need to operate on the `getProvisionedConcurrencyConfig` API. Waiters are modeled directly by the service and since Lambda doesn't include one here we can't provide a method for it.

You'll likely need to write your own logic for polling `getProvisionedConcurrencyConfig` until the response contains `"Status": "READY"` (as opposed to `"IN_PROGRESS"` or `"FAILED"`).

**Reply by ianbotsf** [reply, marked as answer]
Something like this I think:

```kotlin
suspend fun waitUntilConcurrencyUpdated(func: String, alias: String) {
    while (true) {
        val status = lambda.getProvisionedConcurrencyConfig {
            functionName = func
            qualifier = alias
        }

        when (status.status) {
            ProvisionedConcurrencyStatusEnum.Ready -> return
            ProvisionedConcurrencyStatusEnum.Failed -> error("Could not update provisioned concurrency")
            else -> { /* continue loop */ }
        }

        delay(3.seconds) // wait a sensible amount of time to avoid throttling
    }
}
```

**Reply by NikolayMetchev** [reply]
Thanks. I already implemented something very similar. What are the chances a waiter gets added in the future?

**Comment by NikolayMetchev** [comment]
Namely `ProvisionedConcurrencyStatusEnum`

**Comment by RanVaknin** [comment]
Hi @NikolayMetchev ,

I have created a request to add this waiter with the Lambda service team (ticket `#P139856257`). 
We don't know if / when this will get added, but if you have access to AWS Support via the developer console, you can open a support ticket and reference the ticket ID I shared above to try and get an update on this.

Since it seems like you're unblocked for now, and this is not actionable by the SDK team I'm inclined to close this discussion.

Thanks again for bringing this to our attention. 
All the best,
Ran~
